allel/io/vcfreader.py

```diff
@@ -0,0 +1,255 @@
+            t += 1 << ((l << 1) + l)
+            s -= 3
+
+    def _parse_index(self, index):
+        """
+        read and parse the index file, set the virtual offset for the target region,
+        check out https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3042176/
+        and https://samtools.github.io/hts-specs/tabix.pdf
+        for more information.
+
+        Parameters
+        ----------
+        index: str
+            path for the index file
+
+        """
+        with gzip.open(index, 'rb') as f:
+            magic = f.read(4)
+            if magic != b'TBI\x01':  # check magic
+                raise RuntimeError('invalid tabix index file: {}'.format(index))
+            header = np.frombuffer(f.read(4 * 8), dtype=np.int32)
+            if header[1] != 2:  # check format
+                raise RuntimeError('invalid VCF index file: {}'.format(index))
+            # this is the comment marker use to determine VCF header, usually set to b'#'
+            self._meta = chr(header[5]).encode('ascii')
+            try:
+                rid = f.read(header[7]).split(b'\x00').index(self._chrom)
+            except ValueError:
+                # contig name is not exist, read header only
+                self._chunk_begin = -1
+                return
+            # seek to our target index
+            for _ in range(rid):
+                for _ in range(struct.unpack('<i', f.read(4))[0]):  # n_bins
+                    f.seek(4, os.SEEK_CUR)  # bin_key
+                    n_chunk = struct.unpack('<i', f.read(4))[0]  # n_chunk
+                    f.seek(8 * 2 * n_chunk, os.SEEK_CUR)  # cnk_beg, cnk_end
+                n_intv = struct.unpack('<i', f.read(4))[0]  # n_intv
+                f.seek(8 * n_intv, os.SEEK_CUR)  # ioff
+            # read our target index
+            bidx = {bin_key: None for bin_key in self.reg2bins(self._begin, self._end)}
+            for _ in range(struct.unpack('<i', f.read(4))[0]):  # n_bins
+                bin_key = struct.unpack('<I', f.read(4))[0]  # bin
+                n_chunk = struct.unpack('<i', f.read(4))[0]  # n_chunk
+                if bin_key in bidx:
+                    chunks = np.frombuffer(f.read(8 * 2 * n_chunk), dtype=np.uint64)
+                    bidx[bin_key] = chunks.reshape(n_chunk, -1)  # cnk_beg, cnk_end
+                else:
+                    f.seek(8 * 2 * n_chunk, os.SEEK_CUR)  # cnk_beg, cnk_end
+            n_intv = struct.unpack('<i', f.read(4))[0]  # n_intv
+            ioff = np.frombuffer(f.read(8 * n_intv), dtype=np.uint64)  # ioff
+        # linear index: record fall into 16kb(=16 * 2 ** 10) interval
+        min_ioff = ioff[self._begin >> 14]
+        # binning index: record cluster in large interval
+        overlap = np.concatenate([chunks for bin_key, chunks in bidx.items() if chunks is not None])
+        # coupled binning and linear indices, filter out low level bins
+        chunk_begin, *_, chunk_end = np.sort(np.ravel(overlap[overlap[:, 0] >= min_ioff]))
```

Comment:
I'd be very grateful if you could explain what's happening here, for my own understanding. What do we gain by using the binning and linear indices together?

Reply:
Nice question, I am not clear here, too.

So the code here suppose to get the `chunk_end` value, becouse we want to stop reading the VCF as soon as we get through the target region. To do this, in my opinion, we have 3 solution:

1. The best solution: parse and check the record, once we get through the target region, stop reading. But the parse and check process has implement in the fellow process, so that is duplicate. PASS.

2. The easy understanding solution: using the linear index, read until the block which target region end position fall into. But this solution is a bit annoying to me. Becouse the linear index only track the begin offset, to get the end offset, we might need to parse the next chromosome index and/or check wether the block fully contain the record overlap our target region. PASS.

3. The solution here the code using: using the binning index, the binning index track both begin and end offset, so sort and get the biggest offset which is the point we stop reading the VCF.

There is another question related to these code, too. Here, I filter out the lower level bins which might contain records that start before target region but overlap our target region, for example:
```
-------------------------|---query-region---|-----------------
------|---record-will-be-ignore---|-------------------------
```
But if I understand the comment
```
# N.B., still pass the region parameter through so we get strictly only
# variants that start within the requested region. See also
# https://github.com/alimanfoo/vcfnp/issues/54
```
corectly, these records wil be filter out, so the solution here should be ok.

Reply:
Hi @yangmqglobe, thanks so much for following up here, many apologies for slow response. 

This is tricky, I will try to discuss this first:

> There is another question related to these code, too. Here, I filter out the lower level bins which might contain records that start before target region but overlap our target region.

From [Li (2011)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3042176/): "In the linear index, we keep for each tiling 16 kb window the virtual file offset of the leftmost record (i.e. having the smallest start coordinate) that overlaps the window." 

E.g., what if you have a VCF with a large duplication that starts at position 20 kb and ends at 40 kb, and all other records are SNPs. And then you query for variants in the region 35-50 kb. You take the query start (35 kb), floor divide by 16 kb, to get the corresponding window position within linear index, in this case 2, i.e., the third window, which spans 32-48 kb. You then access the third element within the linear index, which gives you the virtual file offset for the first variant overlapping the window 32-48 kb. This will be the virtual file offset for the record containing the large deletion starting at position 20 kb, because it is the first record overlapping the 32-48 kb window. 

In other words, I expect the linear index may give you variants that start before the query region. So filtering out results from the binning index won't stop this from happening. 

But as you also point out, the underlying VCF parsing code in scikit-allel will filter these out when it starts reading, so I don't think we have to worry about this. The code that handles this is [here](https://github.com/cggh/scikit-allel/blob/655254106c09a23ba9c6dc40c83b4a443a66bcfc/allel/opt/io_vcf_read.pyx#L1045). (As an aside, we might want to consider making this behaviour optional, so a user can retrieve all variants overlapping the query region like tabix does, not just those which strictly start within the query region, but that's a separate issue that could be address later.)

Reply:
Regarding this point:

> So the code here suppose to get the chunk_end value, becouse we want to stop reading the VCF as soon as we get through the target region. To do this, in my opinion, we have 3 solution.

If the binning index gives the end offset, then this sounds like a reasonable option to figure out when to stop reading. To be honest I still don't fully understand the binning index, so I cannot make a perfect judgement here.

An alternative I think is to use the linear index to find the offset to start reading, and then just keep reading until the parser finds a record which starts beyond the query end. I.e., let the underlying parser figure out when to stop reading. 

Currently the parser does check whether the current variant record starts beyond the end of the query region, [code here](https://github.com/cggh/scikit-allel/blob/655254106c09a23ba9c6dc40c83b4a443a66bcfc/allel/opt/io_vcf_read.pyx#L1059). However, that code currently is designed for the case where a VCF is not sorted (yes that happens) and so it only skips the variant, it does not terminate parsing. However, we could pass down a piece of information that the VCF is sorted (which we know it will be if querying via tabix index), and the parser could then use this to terminate parsing rather than just skipping the record.  

In other words, I *think* it would be possible to just use the linear index to find a suitable file offset to start reading from, and then let the parser figure out when to stop reading based on parsing records.

Does that fit your understanding? Do you have any thoughts on which strategy is better?

Reply:
Yes, I wonder why there is not a `sorted` parameter all the time.